(#413) Add 420chan support

Lolzen opened the pull request:
- [x] SiteSetup
- [x] read threads/comments
- [x] thumbnails
- [x] posting
- [x] thread pinning
- [ ] ~~savedReply (You)s~~
>current API doesn't make this easy, rather not possible in a reliable way.
- [ ] ~~reply/image count~~
>tried to circumvent the missing API key by manually counting images and passing the value to builder.images, but it either reset on every occourance or added up and didn't reset the counter at all
>~~may research and retry at a later point.~~ replies key is present in threadview mode.
- [X] sage
- [X] reporting
>current implementaion is limited and straight forward. Contrary to 4chan's report, there's no seperate url/window we can use. Therefore the options (reason & note) are predetermined.
- [ ] review text formatting
>should be almost complete. However due to the weird symbiosis by modifying the comments in TaimabaApi and then adding rulesets in TaimabaCommentParser this wil lead to issues and is more a long term process of catching and perfecting all cases
>maybe going the similar route which dvach's implementation of extending the PostParser is a better approach. Will look into that, but for now it's good enough.
- [ ] when a thread 404s it returns an API parse error instead of the 404 message
>absolutely no idea
>https://abload.de/img/screenshot_matlog_202onk2b.png
>https://api.420chan.org/b/res/4952702.json

Lolzen committed:
[WiP] [DNM] Add rudimentary 420chan support

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/sites/Chan420.java:
```diff
@@ -0,0 +1,150 @@
+        setName("420Chan");
+        setIcon(SiteIcon.fromFavicon(HttpUrl.parse("https://420chan.org/favicon.ico")));
+
+        setBoards(
```
Everything is already enumerated in http://api.420chan.org/boards.json; you should implement the boards method when declaring your endpoints class below and make a Chan420BoardsRequest class to parse that json.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaAntispam.java:
```diff
@@ -0,0 +1,105 @@
+/*
```
If this class is exactly the same as VichanAntiSpam, then there's no reason to have it at all. I can already tell that this is old code anyways, as it has an OkHttpClient instance, which should be dependency injected now.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaCommentParser.java:
```diff
@@ -0,0 +1,30 @@
+    public TaimabaCommentParser() {
+        addDefaultRules();		
+		setQuotePattern(Pattern.compile(".*(\\d+)"));
+        setFullQuotePattern(Pattern.compile("/(\\w+)/\\w+/(\\w+)/(\\d+)#(\\d+)"));
```
You said you had issues with formatting not being applied; see VichanCommentParser for how extra parsing rules are added.

Lolzen committed:
change from tabs to spaces

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaEndpoints.java:
```diff
@@ -0,0 +1,85 @@
+
+	//https://cdn.420chan.org:8443/narcbot/ajaxReport.jsp
+    @Override
+    public HttpUrl delete(Post post) {
```
Does 420chan even have a delete option? Not from what I can tell. Seems abuse-able.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaEndpoints.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public HttpUrl reply(Loadable loadable) {
+        return sys.builder().s(loadable.board.code).s("taimaba.pl").url();
```
If you're having problems posting, you might want to see what data actually needs to be sent by looking at what network requests happen during a post, using some browser development feature.

Lolzen replied:
i've gatherd some data with Open Web Sniffer, but i'm pretty clueless how to implement it.
Read some of the things on the Overchan implementation. Maybe i can get some clues.
I don't really grasp it, but i attached the data i gathered below. I'll snoop around if i can find similarities in vichan implementations, i remeber somewhere reading field4 for example.

data when posting:
hostname:  boards.420chan.org
method:		POST
url:		/b/taimaba.pl
		type	main_frame

POST data:
board:	b
fart:	somenumber
field4:	test (comment)
file:	(empty)
link:	(empty)
name:	(empty)
		(empty)
parent:	somenumber (threadId)
password:	generated password
sage:	on (boolean?)
task: 	post
		
response headers:
status:		302
date:		<date>
set-cookie:	pwd[...]
location:	/b/
cf-cache-status:	DYNAMIC
some cloudfare stuff

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaActions.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @Override
+    public void setupPost(Reply reply, MultipartHttpCall call) {
```
The reason why you can't post is probably because a bunch of these methods are constructing the wrong sort of request. You should use a browser to see what network traffic happens during a request so you can perform the right things in these methods.

Lolzen committed:
fetch boards through API

Lolzen committed:
fix compiling

idk how it compiled just minutes ago

Lolzen committed:
remove antispam/delete stuff

Lolzen committed:
prevent app crash due to missing data

taken from dvach

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaActions.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /*@Override
+    public void prepare(MultipartHttpCall call, Reply reply, ReplyResponse replyResponse) {
```
If the code is not used it should be removed.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaActions.java:
```diff
@@ -0,0 +1,122 @@
+
+    @Override
+    public void handlePost(ReplyResponse replyResponse, Response response, String result) {
+        Matcher auth = Pattern.compile("\"captcha\": ?true").matcher(result);
```
Patterns can be made final static so we don't create the every time anew.

Lolzen replied:
i think this section could be simplified as 420chan doesn't use a captcha system, but some kind of anti spam.
https://420chan.org/blog/2013/12/18/captcha-free-anti-spam-tech-now-live/

K1rakishou replied:
If it won't be used then you can just remove it.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaActions.java:
```diff
@@ -0,0 +1,122 @@
+            replyResponse.errorMessage = Jsoup.parse(err.group(1)).body().text();
+        } else {
+            HttpUrl url = response.request().url();
+            Matcher m = Pattern.compile("/\\w+/\\w+/(\\d+)").matcher(url.encodedPath());
```
Patterns can be made final static so we don't create the every time anew.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaActions.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    public Pattern errorPattern() {
+        return Pattern.compile("<h1[^>]*>Error</h1>.*<h2[^>]*>(.*?)</h2>");
```
Patterns can be made final static so we don't create the every time anew.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaApi.java:
```diff
@@ -0,0 +1,374 @@
+            String key = reader.nextName();
+
+            switch (key) {
+				case "no":
```
Would be nice if you could reformat this code (you can do that by just pressing Ctrl+Alt+L in Android Studio)

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaApi.java:
```diff
@@ -0,0 +1,374 @@
+                case "closed":
+                    builder.closed(reader.nextInt() == 1);
+                    break;
+                /*case "closedby":
```
If the code is not used it should be removed.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaApi.java:
```diff
@@ -0,0 +1,374 @@
+                case "trip":
+                    builder.tripcode(reader.nextString());
+                    break;
+                /*case "staff":
```
If the code is not used it should be removed.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaApi.java:
```diff
@@ -0,0 +1,374 @@
+                case "h":
+                    fileHeight = reader.nextInt();
+                    break;
+                /*case "tn_w":
```
If the code is not used it should be removed.

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/sites/chan420/Chan420BoardsRequest.java:
```diff
@@ -0,0 +1,108 @@
+            String key = reader.nextName();
+
+            switch (key) {
+                /*case "category":
```
If the code is not used it should be removed.

K1rakishou reviewed (changes requested):
Just a couple of minor things.

Lolzen committed:
add textformatting rules

Lolzen committed:
reformat

Lolzen committed:
remove unused code

Lolzen committed:
use final static for errorPattern

Lolzen commented:
So, i'm pretty much where i was before except by commenting out defaultRules in TaimabaCommentParser strikethrough is not considered a spoiler. The necessary changes made to TaimabaApi were ripped off from Overchan. That's kinda why i didn't include these in the first place, because now the problem is that whatever i've tried i couldn't manage to properly implement inlinequoting (always weird results when it worked, like ALL text or only the last inline quote colored) and quoting (was linkified but clicking on it had no result).

I'll be focusing on trying to get posting going even though i'll probably need a lot of time.

Lolzen committed:
add inlite quote formatting

Lolzen committed:
update call parameters

still throws an error, don't know how to proceed further

Lolzen committed:
text formatting final touches

*use defaultrules (for quotes to work)
*rename tagrule "s" to "strike" to prevent colliding with defaultrule for spoilers
*use correct quote patterns
*remove comments/cleanup

Lolzen committed:
update TaimabaActions

Lolzen committed:
make postink kinda work

throws an onPostComplete error, but the post arrives

Lolzen committed:
add [sub] formatting

Lolzen committed:
cleanup now unused imports

Lolzen commented:
Posting does work, but throws an onPostComplete error, unsure why. Probably because replyResponse.posted is not true.
I noticed that on Kuroba the inline-quote formatting fucks up reply-quotes (clickable but deformed)
The pattern used should be more precise, but i can't construct a correct one. 

These are the two issues i see as of now.

Adamantcheese commented:
In your Actions class, there's a handlePost method that takes care of setting replyResponse.posted. You may need to adjust the regex inside that method for it to properly set that, so you can do what Wired7 did, which is copy the method and change the regex. The site itself seems to redirect to the board homepage, which is stupid, but will probably make your life more difficult for determining what thread number and post number the reply response is associated with, since those are being parsed from the request URL.

Lolzen committed:
add another text formatting possibility for spoilers

Lolzen commented:
Yep figured that much out, but i couldn't really figure out the correct pattern. However after i read that the site redirects you to the board overview anyways i figured i try a hacky method of just storing the threadNo, postNo & password from reply values of the setupPost method and use them directly. 
Feels pretty hacky, and i don't think it's the correct way to handle things but at least it works without throwing an error now. Anyways the next problem is that your own replies won't be saved, meaning no highlighting and no (You)s consequently. I've looked at onPostComplete in ReplyPresenter and it seems it should be fetched from loadable, as the other values are now passed. Thread pinning does work.

I'll commit the way it is now for review, but i'll keep tinkering around and revert/update once i get the proposed way working.

There also are a lot of minor issues which need to be taken care of/worked around (reply/image count, when posting with an image the file size is a negative value and triggers the warnign "file size too big" (but posts correctly), sage doesn't work) I may find some more, but these are on the todo list for now.

EDIT: correct errorPattern is now used and errors show up correctly.
As said before the redirect doesn't allow us to fetch the postNo from the url, so this needs to somehow be available or made available and used afterwards.

Lolzen committed:
[hack] prevent onPostComplete error

Lolzen committed:
use correct errorPattern regex

Lolzen committed:
show errormessage if encountered

Lolzen committed:
fix inlinequote/quote woes

Lolzen committed:
set correct maxFileSize limits

Lolzen commented:
moved comment to first comment

Lolzen committed:
add another text formatting possibility for italic

Lolzen committed:
remove incorrect postNo reference

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/sites/chan420/Chan420BoardsRequest.java:
```diff
@@ -0,0 +1,118 @@
+                    }
+                    break;
+                case "title":
+                    board.description = reader.nextString();
```
This should be board.name, not board.description; the board.hasMissingInfo check below means that no boards would ever be added for display in the selection window.

Lolzen replied:
ah thanks for catching that. have it correct in my private fork but apparently forgot to change here.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaActions.java:
```diff
@@ -0,0 +1,101 @@
+            call.fileParameter("file", reply.fileName, reply.file);
+        }
+
+        if (reply.options == "sage") {
```
Try to use .equals instead of == for string comparisons.

Lolzen replied:
that worked like a charm. thanks!

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaApi.java:
```diff
@@ -0,0 +1,340 @@
+        queue.addForParse(builder);
+    }
+
+    private PostImage readPostImage(JsonReader reader, Post.Builder builder,
```
For /f/, you may want to choose to do something with SWF files, as they aren't supported in the image viewer.

Lolzen replied:
my idea is using a placeholder thumbnail so you can still download the files.
tested and it works, (currently just using a imagehoster website) altough it will display an error "failed to show image, deep-zooming loading failed"on tapping on it as obviously swf format isn't natively supported on android and it tries to load that.

but i think we could go on from there. there's no static image on 420chan unfortunately as there simply is nothing displayed but a link to the file itself, however fully disabling image display would essentially render the whole board pretty much useless imho.

i'll commit my current solution for further review.

Lolzen replied:
whatever is to be done has to be done for mp3 files as well

Lolzen replied:
i think doing something similar to 462029b is the way to go
I've tested that on my private fork and it's working.

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/sites/chan420/Chan420.java:
```diff
@@ -0,0 +1,113 @@
+
+import okhttp3.HttpUrl;
+
+public class Chan420 extends CommonSite {
```
You're working with an older version of the code, so before merging you'll need to add in ChunkDownloaderSiteProperties and the mediaHosts fields. For ChunkDownloaderSiteProperties, both should be false, as the byte count is wrong and no hash is sent. for mediaHosts you should use boards.420chan.org. Take a look at how those are implemented in the current code, using Sushichan and Chan4 as examples.

Lolzen replied:
commited the changes as requested derived from looking at examples.
tested based on current code (6c52470) and verified working.

i'll leave this open for review just in case.

Lolzen committed:
fix sage

Lolzen committed:
fix wrong builder call

Lolzen committed:
use placeholder thumbnails for swf files

Lolzen committed:
add ChunkDownloaderSiteProperties

Lolzen committed:
fix missing import & use new format for desktopUrl override

Lolzen committed:
update Siteregistry to not conflict with 8kun's addition

Lolzen committed:
add ability to report posts

limitations:
*the webview isn't helpful other than confirming the report has been sent
*the reason will always be "RULE_VIOLATION", never "ILLEGAL_CONTENT"
*the (supposed to be helpful and descriptive text) will always be "Kuroba generated report")

Lolzen committed:
add mp3 thumbnails & use a same styled swf thumbnail icon

these icons are from https://www.pngrepo.com/vectors/files-3/2&type=multicolor

Licensed under Creative Commons 4.0

Lolzen committed:
simplyfy feature override

Adamantcheese commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/common/taimaba/TaimabaEndpoints.java:
```diff
@@ -88,4 +89,16 @@ public HttpUrl pages(Board board) {
+            .addPathSegment("ajaxReport.jsp")
+            .addQueryParameter("postId", String.valueOf(post.no))
+            .addQueryParameter("reason", "RULE_VIOLATION")
+            .addQueryParameter("note", "Kuroba generated report")
```
You might want to avoid using something so easily traceable to this application, in case the admins don't like it and end up banning the user agent that sent the report. Because reports don't have a captcha on them, that can end up being really abused.

Lolzen replied:
should i revert the whole commit, or just edit the note part?

Adamantcheese replied:
Just edit the note part so it's more a generic report. Seems like the note is meant to be an explanation of why it's breaking the rules anyways.

Lolzen committed:
remove note text

Lolzen committed:
use a neutral file thumbnail for audio formats

K1rakishou commented on Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/site/sites/chan420/Chan420.java:
```diff
@@ -0,0 +1,137 @@
+    private static final String TAG = "420Chan";
+
+    public static final CommonSiteUrlHandler URL_HANDLER = new CommonSiteUrlHandler() {
+        private final String[] mediaHosts = new String[]{"boards.20chan.org"};
```
Is it really supposed to be "20chan" here?

Lolzen replied:
good catch!

Lolzen committed:
typo
